### PR TITLE
Issue #118, Support Edm.Date with Entity Framework.

### PR DESCRIPTION
**Pull request by xuzhg**
Issue #118, Support Edm.Date with Entity Framework.

To make date & time work between OData and Entity Framework.

### CODE REVIEW

File: OData/src/System.Web.OData/OData/Builder/EdmTypeBuilder.cs [hunk 1]
```diff
@@ -152,8 +152,10 @@ private void CreateStructuralTypeBody(EdmStructuredType type, StructuralTypeConf
                 switch (property.Kind)
                 {
                     case PropertyKind.Primitive:
-                        PrimitivePropertyConfiguration primitiveProperty = property as PrimitivePropertyConfiguration;
-                        EdmPrimitiveTypeKind typeKind = GetTypeKind(primitiveProperty.PropertyInfo.PropertyType);
+                        PrimitivePropertyConfiguration primitiveProperty = (PrimitivePropertyConfiguration)property;
+                        EdmPrimitiveTypeKind typeKind = primitiveProperty.TargetEdmTypeKind ??
+                                                        GetTypeKind(primitiveProperty.PropertyInfo.PropertyType);
+
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Builder/EdmTypeBuilder.cs]: remove this blank line

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 2]
```diff
@@ -1948,6 +1954,11 @@ private static bool IsDateTime(Type type)
             return IsType<DateTime>(type);
         }
 
+        private static bool IsTimeSpan(Type type)
+        {
+            return IsType<TimeSpan>(type) || IsType<TimeSpan>(type);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: remove second statement?

File: OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs [hunk 3]
```diff
@@ -88,8 +89,24 @@ public static object ConvertPrimitiveValue(object value, Type type)
                         return dateTimeOffsetValue.DateTime;
                     }
 
+                    if (value is Date)
+                    {
+                        Date dt = (Date)value;
+                        return (DateTime)dt;
+                    }
+
                     throw new ValidationException(Error.Format(SRResources.PropertyMustBeDateTimeOffset));
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs]: ErrorMessage should change?

File: OData/src/System.Web.OData/System.Web.OData.csproj [hunk 4]
```diff
@@ -484,5 +486,8 @@
   <ItemGroup>
     <None Include="packages.config" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="OData\Formatter\Converter\" />
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/System.Web.OData.csproj]: nothing in this folder?

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace System.Web.OData.Builder
+{
+    /// <summary>
+    /// Extensions method for see <see cref="PrimitivePropertyConfiguration"/>.
+    /// </summary>
+    public static class PrimitivePropertyConfigurationExtensions
+    {
+        /// <summary>
+        /// If this primitive property is <see cref="System.DateTime"/>, this method will make the target Edm type kind as <see cref="Date"/>
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: wrap the line longer than 120?

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            {
+                throw Error.Argument("property", SRResources.MustBeDateTimeProperty, property.PropertyInfo.Name, property.DeclaringType.FullName);
+            }
+
+            property.TargetEdmTypeKind = EdmPrimitiveTypeKind.Date;
+            return property;
+        }
+
+        /// <summary>
+        ///  If this primitive property is <see cref="System.TimeSpan"/>, this method will make the target Edm type kind as <see cref="TimeOfDay"/>
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: wrap the line longer than 120?

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        public static PrimitivePropertyConfiguration AsTimeOfDay(this PrimitivePropertyConfiguration property)
+        {
+            if (property == null)
+            {
+                throw Error.ArgumentNull("property");
+            }
+
+            if (!TypeHelper.IsTimeSpan(property.RelatedClrType))
+            {
+                throw Error.Argument("property", SRResources.MustBeTimeSpanProperty, property.PropertyInfo.Name, property.DeclaringType.FullName);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: wrap the line longer than 120?

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        public static PrimitivePropertyConfiguration AsDate(this PrimitivePropertyConfiguration property)
+        {
+            if (property == null)
+            {
+                throw Error.ArgumentNull("property");
+            }
+
+            if (!TypeHelper.IsDateTime(property.RelatedClrType))
+            {
+                throw Error.Argument("property", SRResources.MustBeDateTimeProperty, property.PropertyInfo.Name, property.DeclaringType.FullName);
```
Comment (VikingsFan) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: wrap the line longer than 120?

File: OData/src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                throw Error.ArgumentNull("edmProperty");
+            }
+
+            if (edmProperty.AddedExplicitly)
+            {
+                return;
+            }
+
+            PrimitivePropertyConfiguration primitiveProperty = edmProperty as PrimitivePropertyConfiguration;
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: `var` for new or edited code.

File: OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs [hunk 10]
```diff
@@ -88,8 +89,24 @@ public static object ConvertPrimitiveValue(object value, Type type)
                         return dateTimeOffsetValue.DateTime;
                     }
 
+                    if (value is Date)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs]: `value as Date`, and then use a null check.

File: OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs [hunk 11]
```diff
@@ -88,8 +89,24 @@ public static object ConvertPrimitiveValue(object value, Type type)
...
+                    {
+                        Date dt = (Date)value;
+                        return (DateTime)dt;
+                    }
+
                     throw new ValidationException(Error.Format(SRResources.PropertyMustBeDateTimeOffset));
                 }
+                else if (type == typeof(TimeSpan))
+                {
+                    if (value is TimeOfDay)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs]: prefer `as` than `is`.

File: OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs (hunk 2 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: Maybe IsType<TimeSpan?>(type)..

File: OData/src/System.Web.OData/System.Web.OData.csproj (hunk 4 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/System.Web.OData.csproj]: removed.

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs (hunk 8 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: done.

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs (hunk 7 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: done

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs (hunk 6 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: done

File: OData/src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs (hunk 9 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Builder/Conventions/Attributes/ColumnAttributeEdmPropertyConvention.cs]: ok

File: OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs (hunk 10 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs]: Date is a `struct`, can't use to do null check.

File: OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs (hunk 11 above)
Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Formatter/EdmPrimitiveHelpers.cs]: `TimeOfDay` is a `struct`, so, have to use `is`

**Comment by xuzhg** [comment]
Update. Thanks.

File: OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    /// </summary>
+    public static class PrimitivePropertyConfigurationExtensions
+    {
+        /// <summary>
+        /// If this primitive property is <see cref="System.DateTime"/>, this method will make the target
+        /// Edm type kind as <see cref="Date"/>
+        /// </summary>
+        /// <param name="property">Reference to the calling primitive property configuration.</param>
+        /// <returns>Returns itself so that multiple calls can be chained.</returns>
+        public static PrimitivePropertyConfiguration AsDate(this PrimitivePropertyConfiguration property)
```
Comment (congysu) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: The api does not look correct.  A primitive property configuration can be anything.  We should enable user to write correctly easier, not another way around - with run time exceptions.

Comment (xuzhg) [inline OData/src/System.Web.OData/OData/Builder/PrimitivePropertyConfigurationExtensions.cs]: Yesterday review meeting, we discussed it and it makes sense to developer to use these extension methods.
Thanks.

**Comment by xuzhg** [comment]
The corresponding spec can be find [here](http://odata.github.io/WebApi/#12-01-DateAndTimeOfDayWithEF)

**Comment by xuzhg** [comment]
Merged. 1221b2e

**Pull request closed without merging**